rasterio/__init__.py

```diff
@@ -190,46 +189,60 @@ def open(fp, mode='r', driver=None, width=None, height=None, count=None,
             "Blacklisted: file cannot be opened by "
             "driver '{0}' in '{1}' mode".format(driver, mode))
 
-    # Special case for file object argument.
+    # If the fp argument is a file-like object and can be adapted by
+    # rasterio's FilePath we do so. Otherwise, we use a MemoryFile to
+    # hold fp's contents and store that in an ExitStack attached to the
+    # dataset object that we will return. When a dataset's close method
+    # is called, this ExitStack will be unwound and the MemoryFile's
+    # storage will be cleaned up.
     if mode == 'r' and hasattr(fp, 'read'):
-
-        @contextmanager
-        def fp_reader(fp):
-            if have_vsi_plugin:
-                vsi_file = FilePath(fp)
-            else:
-                vsi_file = MemoryFile(fp.read())
-            dataset = vsi_file.open(driver=driver, sharing=sharing)
-            try:
-                yield dataset
-            finally:
-                dataset.close()
-                vsi_file.close()
-
-        return fp_reader(fp)
+        if have_vsi_plugin:
+            return FilePath(fp).open(driver=driver, sharing=sharing)
+        else:
+            memfile = MemoryFile(fp.read())
+            dataset = memfile.open(driver=driver, sharing=sharing)
+            ctxstack = ExitStack()
+            ctxstack.enter_context(env_ctx_if_needed())
+            ctxstack.enter_context(memfile)
+            dataset._env = ctxstack
+            return dataset
 
     elif mode in ('w', 'w+') and hasattr(fp, 'write'):
-
-        @contextmanager
-        def fp_writer(fp):
-            memfile = MemoryFile()
-            dataset = memfile.open(driver=driver, width=width, height=height,
-                                   count=count, crs=crs, transform=transform,
-                                   dtype=dtype, nodata=nodata, sharing=sharing, **kwargs)
-            try:
-                yield dataset
-            finally:
-                dataset.close()
-                memfile.seek(0)
-                fp.write(memfile.read())
-                memfile.close()
-
-        return fp_writer(fp)
+        memfile = MemoryFile()
+        dataset = memfile.open(
+            driver=driver,
+            width=width,
+            height=height,
+            count=count,
+            crs=crs,
+            transform=transform,
+            dtype=dtype,
+            nodata=nodata,
+            sharing=sharing,
+            **kwargs
+        )
+        ctxstack = ExitStack()
+        ctxstack.enter_context(env_ctx_if_needed())
+        ctxstack.enter_context(memfile)
```

Comment:
should `dataset` be also registered in ctxstack?

Reply:
No, only the objects which are named only in the scope of rasterio.open need callbacks on the stack.